### PR TITLE
Jit IR

### PR DESCRIPTION
**Pull request by hrydgard**
Initial implementation of an "IR"  (intermediate representation) for the recompiler. May allow some advanced inlining and reordering optimizations in the JIT over time.

Does not seem to affect performance much if at all, at the moment.

Tested on ARM and x86/x64. Want this in before I start working on the ARM64 JIT.

### CODE REVIEW

**Comment by unknownbrackets** [comment]
One important thing: I think we should avoid doing any reordering whatsoever if there are any breakpoints that exist, or at least if any are within the block (I think we have a func for that.)

-[Unknown]

**Comment by hrydgard** [comment]
@unknownbrackets , yeah, makes sense.

**Comment by unknownbrackets** [comment]
Here's a rebased version:
https://github.com/hrydgard/ppsspp/compare/master...unknownbrackets:jit-ir

I did not do anything other than make sure it builds, and I'm sure arm64 won't build, but it may be otherwise broken.  There's more distance to improve but having a rebased version helps.  Should I push it over this branch?

-[Unknown]

**Comment by hrydgard** [comment]
@unknownbrackets  yeah that would be good, thanks!

**Comment by unknownbrackets** [comment]
Okay, done - 6ee4d8f...d7583f5.

-[Unknown]

File: Core/MIPS/IR.cpp [hunk 1]
```diff
@@ -0,0 +1,397 @@
...
+		if (e.info & DELAYSLOT) {
+			// Check if replaceable JAL. If not, bail in 2 instructions.
+			bool replacableJal = false;
+			if (e.info & IS_JUMP) {
+				if ((e.op >> 26) == 3) {
+					// Definitely a JAL
+					const ReplacementTableEntry *entry;
+					u32 funcSize;
+					if (CanReplaceJalTo(MIPSCodeUtils::GetJumpTarget(e.origAddress), &entry, &funcSize))
+						replacableJal = true;
```
Comment (unknownbrackets) [inline Core/MIPS/IR.cpp]: Note for self: optimization we could apply in the IR for #8725.

-[Unknown]

File: Core/MIPS/IR.cpp [hunk 2]
```diff
@@ -0,0 +1,397 @@
...
+}
+
+static bool Reorder(IRBlock *block) {
+	bool changed = false;
+
+	// TODO: We only do some really safe optimizations now. Can't do fun stuff like hoisting loads/stores until we have unfolded all branch delay slots!
+	// Well, maybe we could do some of it, but let's not..
+
+	// Sweep downwards
+	for (int i = 0; i < (int)block->entries.size() - 1; i++) {
```
Comment (unknownbrackets) [inline Core/MIPS/IR.cpp]: Note for self: optimization we could apply in the IR for #8725.

-[Unknown]

File: Core/MIPS/x86/CompALU.cpp [hunk 3]
```diff
@@ -61,10 +61,15 @@ namespace MIPSComp
 		u32 uimm = (u16)(op & 0xFFFF);
 		MIPSGPReg rt = _RT;
 		MIPSGPReg rs = _RS;
-		gpr.Lock(rt, rs);
-		gpr.MapReg(rt, rt == rs, true);
-		if (rt != rs)
-			MOV(32, gpr.R(rt), gpr.R(rs));
+		if (gpr.R(rs).IsSimpleReg() && !GetIREntry().IsGPRAlive(rs)) {
+			// NOTICE_LOG(JIT,"immlogic remap at %08x : %08x", js.blockStart, GetCompilerPC());
+			gpr.FlushRemap(rs, rt, GetIREntry().IsGPRClobbered(rs));
```
Comment (unknownbrackets) [inline Core/MIPS/x86/CompALU.cpp]: Note for self: could do this, or could just move to SSA before jitting, or something.

-[Unknown]

File: Core/MIPS/x86/CompLoadStore.cpp [hunk 4]
```diff
@@ -308,7 +308,18 @@ namespace MIPSComp {
 			break;
 
 		case 32: //R(rt) = (u32)(s32)(s8) ReadMem8 (addr); break; //lb
-			CompITypeMemRead(op, 8, &XEmitter::MOVSX, safeMemFuncs.readU8);
+			{
+				IREntry &next = irblock.entries[js.irBlockPos + 1];
+				if (MIPSAnalyst::IsANDIInstr(next.op) && _RT == MIPS_GET_RT(next.op) && MIPS_GET_IMM16(next.op) == 0xFF) {
+					// Seen, but not very often.
+					// TODO: Move to an IR postprocess cleanly as an exercise :)
+					NOTICE_LOG(JIT, "Found lb+andi! Turning lb into lbu  %08x", irblock.address);
```
Comment (unknownbrackets) [inline Core/MIPS/x86/CompLoadStore.cpp]: Note for self: optimization we could apply in the IR for #8725.

-[Unknown]

File: Core/MIPS/x86/CompVFPU.cpp [hunk 5]
```diff
@@ -2455,12 +2455,17 @@ void Jit::Comp_VMatrixInit(MIPSOpcode op) {
 
 	MatrixSize sz = GetMtxSize(op);
 	int n = GetMatrixSide(sz);
+	int vd = _VD;
+	if (n == 4) {
+		// Just remove the transposed-ness. All modes are transpose-invariant.
+		vd &= ~0x20;
```
Comment (unknownbrackets) [inline Core/MIPS/x86/CompVFPU.cpp]: Hmm, maybe should bring this over.

-[Unknown]

**Comment by unknownbrackets** [comment]
At this point, #8725 is probably a more ideal path.  I marked some things we might want to bring over... we should probably close this after merging #8725 though.

-[Unknown]

**Comment by hrydgard** [comment]
Superseded by #8725

**Pull request closed without merging**